torch/csrc/autograd/python_variable.cpp

```diff
@@ -413,6 +427,189 @@ static PyObject * THPVariable_device(THPVariable* self) {
+        }
+        out_no = PyTuple_GET_SIZE(out_args);
+        for (int i = 0; i < out_no; ++i){
+            obj = PyTuple_GET_ITEM(out_args, i);
+            if (PyObject_IsInstance(obj, base)){
+                auto& var = ((THPVariable*)obj)->cdata;
+                var.detach_();
+                new_obj = torch::utils::tensor_to_numpy(var.data());
+                Py_DECREF(obj);
+                PyTuple_SET_ITEM(out_args, i, new_obj);
+            }
+        }
+        PyDict_SetItem(kwargs, key, out_args);
+    }
+    else{
+        out_args = PyTuple_New(((PyUFuncObject*)ufunc)->nout);
+        out_no = PyTuple_GET_SIZE(out_args);
+    }
+    Py_DECREF(key);
+
+    // Call ufunc with np.ndarray types
+    ufunc_method = PyObject_GetAttr(ufunc, method_name);
+    if (ufunc_method == NULL)
+        goto cleanup;
+    result = PyObject_Call(ufunc_method, in_args, kwargs);
+    Py_DECREF(ufunc_method);
+
+    // Result Error Checking
+    if (PyUnicode_CompareWithASCIIString(method_name, "at") == 0){
+        goto cleanup;
+    }
+    if (result == NULL || result == Py_NotImplemented)
+        goto cleanup;
+
+    // Up-Cast Results to ptype
+    if (((PyUFuncObject*)ufunc)->nout == 1){
+        if (PyTypeNum_ISBOOL(PyArray_DESCR((PyArrayObject*)result)->type_num)){
+            //Workaround, torch has no built-in bool tensor
+            new_obj = PyArray_CastToType((PyArrayObject*)result, PyArray_DescrFromType(NPY_UINT8), /*is_fortran=*/false);
+            Py_DECREF(result);
+            result = new_obj;
+        }
+        results = PyTuple_Pack(1, result);
+    }
+    else{
+        results = result;
+        Py_INCREF(results);
+    }
+    Py_DECREF(result);
+    for (int i = 0; i < out_no; ++i){
+        obj = PyTuple_GET_ITEM(out_args, i);
+        if(obj == NULL){
+            obj = PyTuple_GET_ITEM(results, i);
+        }
+        auto var = torch::utils::tensor_from_numpy(obj);
+        result_requires_grad &= at::is_floating_point(var);
+        new_obj = THPVariable_Wrap(make_variable(var));
+        Py_DECREF(obj);
+        obj = new_obj;
+        result_args = PyTuple_Pack(2, (PyObject*)ptype, obj);
+        result_kwargs = PyDict_New();
+        key = PyUnicode_FromString("require_grad");
+        PyDict_SetItem(result_kwargs, key, torch::autograd::utils::wrap(result_requires_grad));
+        new_obj = THPVariable_make_subclass(obj, result_args, result_kwargs);
+        Py_DECREF(obj);
+        Py_DECREF(key);
+        Py_DECREF(result_args);
+        Py_DECREF(result_kwargs);
+        if (new_obj == NULL){
+            result = NULL;
+            goto cleanup;
+        }
+        PyTuple_SET_ITEM(results, i, new_obj);
+    }
+    if (PyTuple_GET_SIZE(results) == 1){
+        result = PyTuple_GET_ITEM(results, 0);
+    }
+    else{
+        result = results;
+    }
+    Py_INCREF(result);
+    Py_DECREF(results);
+
+cleanup:
+    Py_DECREF(in_args);
+    Py_DECREF(out_args);
+    Py_DECREF(ptype);
+    Py_DECREF(base);
```

Comment:
Explicit refcounting is giving me hives. We have C++ helper classes for doing RAII on these, please use them!

Reply:
I have cleaned up this function so that it uses pybind11 whenever possible.